File: day_15_shorter.R
MyFile <- "data/input_15.txt"
data.source <- readLines(MyFile)
n <- max(nchar(data.source))
data<- read.fwf(textConnection(data.source), rep(1, n)) 
######### 
partB <- FALSE  # part A or B
if (partB==TRUE){
  newdata <- cbind(data,data+1,data+2,data+3,data+4)
  newdata <- rbind(newdata,newdata+1,newdata+2,newdata+3,newdata+4)
  newdata <-  newdata %% 9
  newdata[newdata==0] <- 9
  data <- newdata
  n <- 5*n
}
###########
nn <- n*n
start=Sys.time()
# transformed to cpprouting problem
library(cppRouting)
library(igraph)
graph<-data.frame(from=NA, to=NA, weight=NA)

for (i in 1:n){
  graph <- rbind(graph, c(from=(i-1)*n+2,to=(i-1)*n+1,weight=data[i,1]),
                 c(from=(i-1)*n+1+n,to=(i-1)*n+1,weight=data[i,1]),
                 c(from=(i-1)*n+1-n,to=(i-1)*n+1,weight=data[i,1]))
  for (j in 2:(n-1)){
    graph <- rbind(graph, c(from=(i-1)*n+j-1,to=(i-1)*n+j,weight=data[i,j]),
    c(from=(i-1)*n+j+1,to=(i-1)*n+j,weight=data[i,j]),
    c(from=(i-1)*n+j+n,to=(i-1)*n+j,weight=data[i,j]),
    c(from=(i-1)*n+j-n,to=(i-1)*n+j,weight=data[i,j]))
  }
  graph <- rbind(graph, c(from=(i-1)*n+n-1,to=(i-1)*n+n,weight=data[i,n]),
                 c(from=(i-1)*n+n+n,to=(i-1)*n+j,weight=data[i,n]),
                 c(from=(i-1)*n,to=(i-1)*n+n,weight=data[i,n]))

}
graph <- na.omit(graph)

graph <- graph[graph$from %in% 1:nn,]
graph <- graph[graph$to %in% 1:nn,]

cpp_graph<-makegraph(graph,directed=TRUE)
get_distance_pair(cpp_graph,from=1,to=nn)
Sys.time()-start

File: day_15_shorter.py
import time
start = time.time()
data = []
f = open("data/input_15.txt",'r')
for line in f:
    x = [int(i) for i in line.rstrip()]
    data.append(x)
n=len(data[0])
nn=n*n

graph = [[]]
for i in range(n):
    # j=0 
    graph.append([i*n+    +2,    i*n+1,data[i][0]]) #from right
    graph.append([(i-1)*n+ 1,    i*n+1,data[i][0]]) #from up
    graph.append([(i+1)*n+ 1,    i*n+1,data[i][0]]) #from down
    for j in range(1,n-1):
        graph.append([i*n+    (j-1)+1,    i*n+j+1,data[i][j]]) #from left
        graph.append([i*n+    (j+1)+1,    i*n+j+1,data[i][j]]) #from right
        graph.append([(i-1)*n+ j+1,       i*n+j+1,data[i][j]]) #from up
        graph.append([(i+i)*n+ j+1,       i*n+j+1,data[i][j]]) #from down
    #j=n-1
    graph.append([i*n+n+1,i*n+n,data[i][n-1]]) #from right
    graph.append([(i-1)*n+ n,      i*n+n,data[i][n-1]]) #from down
    graph.append([(i+1)*n+ n,      i*n+n,data[i][n-1]]) #from up

print(time.time() - start)
